Add richer structure for Stable MIR Projections

klinvill opened the pull request:
Resolves https://github.com/rust-lang/project-stable-mir/issues/49. 

Projections in Stable MIR are currently just strings. This PR replaces that representation with a richer structure, namely projections become vectors of `ProjectionElem`s, just as in MIR. The `ProjectionElem` enum is heavily based off of the MIR `ProjectionElem`.

This PR is a draft since there are several outstanding issues to resolve, including:

- How should `UserTypeProjection`s be represented in Stable MIR? In MIR, the projections are just a vector of `ProjectionElem<(),()>`, meaning `ProjectionElem`s that don't have Local or Type arguments (for `Index`, `Field`, etc. objects). Should `UserTypeProjection`s be represented this way in Stable MIR as well? Or is there a more user-friendly representation that wouldn't drag along all the `ProjectionElem` variants that presumably can't appear?
- What is the expected behavior of a `Place`'s `ty` function? Should it resolve down the chain of projections so that something like `*_1.f` would return the type referenced by field `f`?
- Tests should be added for `UserTypeProjection`

klinvill committed:
Add stable MIR Projections support based on MIR structure

This commit includes richer projections for both Places and
UserTypeProjections. However, the tests only touch on Places. There are
also outstanding TODOs regarding how projections should be resolved to
produce Place types, and regarding if UserTypeProjections should just
contain ProjectionElem<(),()> objects as in MIR.

klinvill commented:
r? @celinval

celinval commented:
@klinvill I'm not an official reviewer, I'll take a look at it though.

r? @spastorino

can you take a look at it please?

celinval reviewed (commented):
I would like to suggest that you crate a new test instead of extending the existing one. Thanks for doing this though!

klinvill committed:
Move SMIR projections tests to new file

klinvill committed:
Replace match assertions against empty slices with is_empty assertions

Asserting is_empty is slightly more concise.

klinvill committed:
Remove rich UserTypeProjection projections in SMIR

It's not clear to me (klinvill) that UserTypeProjections are produced
anymore with the removal of type ascriptions as per
https://github.com/rust-lang/rfcs/pull/3307. Furthermore, it's not clear
to me which variants of ProjectionElem could appear in such projections.
For these reasons, I'm reverting projections in UserTypeProjections to
simple strings until I can get more clarity on UserTypeProjections.

klinvill committed:
Defer Place ty implementation in Stable Mir to later PR

klinvill commented on compiler/stable_mir/src/mir/body.rs:
```diff
@@ -533,6 +644,8 @@ impl Constant {
 }
 
 impl Place {
+    // TODO(klinvill): What is the expected behavior of this function? Should it resolve down the
```
Sounds good, I'll defer this to a separate PR

klinvill committed:
Add SMIR visitor for Places and projections

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6828457936/job/18572620233) [(plain)](https://github.com/rust-lang/rust/commit/d517a1cbda49309c8364daaedfee040371e616fb/checks/18572620233/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=klinvill
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_7eadd066-cf4f-47b3-afff-2e6f028d0692
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=smir-projections
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_7eadd066-cf4f-47b3-afff-2e6f028d0692
GITHUB_REF=refs/pull/117517/merge
GITHUB_REF_NAME=117517/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=5912bf32651e3cc5bc4eedfa2c36a3786cc77a47
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_7eadd066-cf4f-47b3-afff-2e6f028d0692
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_7eadd066-cf4f-47b3-afff-2e6f028d0692
GITHUB_TRIGGERING_ACTOR=klinvill
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117517/merge
GITHUB_WORKFLOW_SHA=5912bf32651e3cc5bc4eedfa2c36a3786cc77a47
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container 726408fdbdbf
 ---> 812d01539d64
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> a5e9e8d79cb5
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-4d2nb9_1/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
    Can't uninstall 'setuptools'. No files were found to uninstall.
Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Collecting virtualenv
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 63.6 MB/s eta 0:00:00
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.13.1 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container 9bfeb028afd2
 ---> c62758507b2c
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> b3eeb244c7c8
 ---> b3eeb244c7c8
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> c2dc23bd440c
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 4ffad5103a43
 ---> 80f8930475b9
Successfully built 80f8930475b9
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:80f8930475b9d60b8299eee41ca6a8db94c25428fbd08562652afdbaead58173
Uploading finished image sha256:80f8930475b9d60b8299eee41ca6a8db94c25428fbd08562652afdbaead58173 to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
80f8930475b9   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
b3eeb244c7c8   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
c62758507b2c   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.8MB    
a5e9e8d79cb5   9 seconds ago    /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
812d01539d64   10 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      5 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      5 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      5 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f1e3339f150>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:80f8930475b9d60b8299eee41ca6a8db94c25428fbd08562652afdbaead58173
---
DirectMap4k:      188352 kB
DirectMap2M:     8200192 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/d4c86cfc4984293d642e5b4f7eece60a6e83e264/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-d4c86cfc4984293d642e5b4f7eece60a6e83e264-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/stable_mir/src/mir/body.rs:469: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (23.3.1)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 39.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 39.1 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 133.2 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:01:00
  local time: Fri Nov 10 18:40:41 UTC 2023
  network time: Fri, 10 Nov 2023 18:40:41 GMT
```

</details>

klinvill committed:
Ignore FieldIdx and VariantIdx examples in docs

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6830906526/job/18579588057) [(plain)](https://github.com/rust-lang/rust/commit/ae1726bfceffa271131d9ef852c1d553688aa6a4/checks/18579588057/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=klinvill
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_b02cba53-f907-43dd-be94-ffc335fb9172
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=smir-projections
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_b02cba53-f907-43dd-be94-ffc335fb9172
GITHUB_REF=refs/pull/117517/merge
GITHUB_REF_NAME=117517/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=8659c488bb089a359e43267e26bcb6d9f993b52a
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_b02cba53-f907-43dd-be94-ffc335fb9172
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_b02cba53-f907-43dd-be94-ffc335fb9172
GITHUB_TRIGGERING_ACTOR=klinvill
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/117517/merge
GITHUB_WORKFLOW_SHA=8659c488bb089a359e43267e26bcb6d9f993b52a
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Removing intermediate container edeabbc6ffe4
 ---> b7b97bda7420
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 52e2af4673d2
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-yd7irzfl/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
  Downloading virtualenv-20.24.6-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 62.1 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.11.0-py3-none-any.whl (17 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 124.0 MB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.13.1 platformdirs-3.11.0 virtualenv-20.24.6
Removing intermediate container 5c047fe025a3
 ---> 079f9575d333
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 7a40d32fecd7
 ---> 7a40d32fecd7
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> c61f73699dad
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container e00e8b8ef9be
 ---> e48c13c5ff4a
Successfully built e48c13c5ff4a
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:e48c13c5ff4a0e1e1beda1dcec601dd0b596295450c0ffebe180f9f0f2414697
Uploading finished image sha256:e48c13c5ff4a0e1e1beda1dcec601dd0b596295450c0ffebe180f9f0f2414697 to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
e48c13c5ff4a   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
7a40d32fecd7   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
079f9575d333   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.8MB    
52e2af4673d2   9 seconds ago    /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
b7b97bda7420   10 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      5 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      5 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      5 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f3ea3e1f6d0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:e48c13c5ff4a0e1e1beda1dcec601dd0b596295450c0ffebe180f9f0f2414697
---
DirectMap4k:      176064 kB
DirectMap2M:     7163904 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/edf0b1db0a7f29d71ee82cfc53bdc170fe74e501/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-edf0b1db0a7f29d71ee82cfc53bdc170fe74e501-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/stable_mir/src/mir/body.rs:469: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (23.3.1)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 29.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 29.8 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 139.1 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
some tidy checks failed
Build completed unsuccessfully in 0:01:01
  local time: Sat Nov 11 00:27:12 UTC 2023
  network time: Sat, 11 Nov 2023 00:27:12 GMT
```

</details>

ouz-a commented:
> 
>     * Within `ProjectionElem`, the `Downcast` variant in MIR includes an `Option<Symbol>` argument that can contain the name of the variant and comments state that it is used for printing the MIR. However, I believe this information can be obtained just using the `VariantIdx` argument. I've therefore omitted the `Option<Symbol>` argument. Is there instead a reason to keep it around?
> 
>     * I haven't been able to generate MIR that contains the `UserTypeProjection` struct. Is it still possible to generate `UserTypeProjection` MIR after type ascriptions were removed in accordance with [De-RFC: Remove type ascription rfcs#3307](https://github.com/rust-lang/rfcs/pull/3307)? If so, in what cases is this struct generated? In the meantime I've left the projections field in `UserTypeProjection` unimplemented (specifically it's still a string).
> 
>     * In MIR, the projections in `UserTypeProjection` and `Place` use different instantiations of the same generic enum `ProjectionElem`. In particular, it seems like `UserTypeProjection` projections only support a subset of the variants. I've therefore opted to hardcode `ProjectionElem` to the instantiation for `Place`, and will create a separate enum if `UserTypeProjection`s should still be included in Stable MIR.
> 
> 

I will try to address these.

1. Imo we should always try to have printable version of a thing available for error reporting and debug printing also this will be useful for https://github.com/rust-lang/rust/pull/117745 once it lands
2/3. I think we can leave them for now and if we hit `UserTypeProjection` we know we have a case for it and we can implement it. This can be done randomly testing stuff in `ui`

celinval commented:
Hi @klinvill, I think it makes sense to remove the `Option<Symbol>` as long as we provide an utility function that retrieves that information. Otherwise, we might want to keep it as pointed out by @ouz-a.

Can you change the `projection` field of `UserTypeProjection` to `Opaque` type for now? Thanks

klinvill commented:
Thanks for the suggestions @ouz-a and @celinval. The inconsistency of having `Option<Symbol>` only to go along with the `VariantIdx` in `Downcast`, and not having a similar field to go with `FieldIdx` in `Field` bugs me a bit. I think it would be better long-term to be consistent with either passing around the additional information, or having that information accessible through a lookup based on the index.

I'd advocate for having that information accessible through a lookup based on the index rather than passing around additional information (which would potentially need to be updated as Variants and Fields evolve). However, such a lookup is currently neither possible (due to `AdtDef` not being implemented) nor convenient (due to `ty` not being implemented for `Place`). Currently, looking up the variant in a projection element would require accessing the local field in the parent `Place` object, then using that (index) to locate the appropriate `LocalDecl`, then following down the projection elements that precede the `Downcast` to get the appropriate `AdtDef`, then retrieving the proper `VariantDef` from the variants in the `AdtDef` (which currently isn't implemented). Ideally, we could instead create a copy of the `Place` definition with the projection elements truncated to either the desired `Downcast` projection or its parent, and then just call the `ty` function to get the appropriate `AdtDef`, but the `ty` function also isn't implemented yet.

So I suppose this is my long-winded way of proposing to drop the `Option<Symbol>` field with the intention of having utility functions that can do the lookup based on the index. However, since it's currently not possible to do this, I propose deferring the utility functions (at least for variant lookups for `Downcast`) until `AdtDef` and `Place::ty` are implemented. Does that sound reasonable to you both @ouz-a and @celinval?

celinval commented:
Sounds reasonable to me. Can you please capture your comment above into an issue in [project-stable-mir](https://github.com/rust-lang/project-stable-mir)? We should revisit this in the near term, and create a follow up PR.

Thank you

klinvill committed:
Make UserTypeProjection projections Opaque

Also shifts comments explaining why Stable MIR drops an optional variant
name field, for `Downcast` projection elements, to the `Place::stable`
function.

celinval commented:
@bors r+ rollup

bors commented:
:pushpin: Commit c036a10ed501cd2c4b501af03c920af3f28d360f has been approved by `celinval`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=celinval c036a10ed501cd2c4b501af03c920af3f28d360f -->
<!-- homu: {"type":"Approved","sha":"c036a10ed501cd2c4b501af03c920af3f28d360f","approver":"celinval","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c036a10ed501cd2c4b501af03c920af3f28d360f with merge eeb892e43bc1d7315c398bcbead8dc11fe529ac5...
<!-- homu: {"type":"BuildStarted","head_sha":"c036a10ed501cd2c4b501af03c920af3f28d360f","merge_sha":"eeb892e43bc1d7315c398bcbead8dc11fe529ac5"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6873492804/job/18693603691) [(plain)](https://github.com/rust-lang-ci/rust/commit/eeb892e43bc1d7315c398bcbead8dc11fe529ac5/checks/18693603691/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6873492804/job/18693671962)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6873492804/job/18693671962","builder_name":"checks-actions"} -->

ouz-a commented:
This is probably nothing 
@bors r+ rollup

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.

bors commented:
:pushpin: Commit c036a10ed501cd2c4b501af03c920af3f28d360f has been approved by `ouz-a`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ouz-a c036a10ed501cd2c4b501af03c920af3f28d360f -->
<!-- homu: {"type":"Approved","sha":"c036a10ed501cd2c4b501af03c920af3f28d360f","approver":"ouz-a","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c036a10ed501cd2c4b501af03c920af3f28d360f with merge 9af957e693f79301ceac4137c7989e9e0a245d90...
<!-- homu: {"type":"BuildStarted","head_sha":"c036a10ed501cd2c4b501af03c920af3f28d360f","merge_sha":"9af957e693f79301ceac4137c7989e9e0a245d90"} -->

ouz-a commented:
Maybe we should re-try this again before rolling it up
@bors r-

ouz-a commented:
@bors r+

bors commented:
:pushpin: Commit c036a10ed501cd2c4b501af03c920af3f28d360f has been approved by `ouz-a`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ouz-a c036a10ed501cd2c4b501af03c920af3f28d360f -->
<!-- homu: {"type":"Approved","sha":"c036a10ed501cd2c4b501af03c920af3f28d360f","approver":"ouz-a","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c036a10ed501cd2c4b501af03c920af3f28d360f with merge 698fcc8219e6dd690b148a23af10a0e5747621fe...
<!-- homu: {"type":"BuildStarted","head_sha":"c036a10ed501cd2c4b501af03c920af3f28d360f","merge_sha":"698fcc8219e6dd690b148a23af10a0e5747621fe"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6873559643/job/18693865308) [(plain)](https://github.com/rust-lang-ci/rust/commit/9af957e693f79301ceac4137c7989e9e0a245d90/checks/18693865308/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6873584812/job/18696207188)
Approved by: ouz-a
Pushing 698fcc8219e6dd690b148a23af10a0e5747621fe to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ouz-a","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6873584812/job/18696207188"},"merge_sha":"698fcc8219e6dd690b148a23af10a0e5747621fe"} -->

rust-timer commented:
Finished benchmarking commit (698fcc8219e6dd690b148a23af10a0e5747621fe): [comparison URL](https://perf.rust-lang.org/compare.html?start=cc8681b64b6f085bb64c0bbbe6739789d3b1eecf&end=698fcc8219e6dd690b148a23af10a0e5747621fe&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=cc8681b64b6f085bb64c0bbbe6739789d3b1eecf&end=698fcc8219e6dd690b148a23af10a0e5747621fe&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=cc8681b64b6f085bb64c0bbbe6739789d3b1eecf&end=698fcc8219e6dd690b148a23af10a0e5747621fe&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.2%  | [3.8%, 4.5%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.7%, -0.5%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-2.5%, -2.5%] | 1     |
| All ❌✅ (primary)                 | -0.6% | [-0.7%, -0.5%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=cc8681b64b6f085bb64c0bbbe6739789d3b1eecf&end=698fcc8219e6dd690b148a23af10a0e5747621fe&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4% | [0.4%, 0.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.4% | [0.4%, 0.4%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=cc8681b64b6f085bb64c0bbbe6739789d3b1eecf&end=698fcc8219e6dd690b148a23af10a0e5747621fe&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 674.399s -> 674.6s (0.03%)
**Artifact size**: 311.08 MiB -> 311.10 MiB (0.00%)
